go/host/host.go

```diff
@@ -659,49 +659,58 @@ func (a *Node) processSharedSecretResponse(_ *ethadapter.L1RespondSecretTx) erro
 	// only process blocks if the node is running
 	for atomic.LoadInt32(a.stopNodeInterrupt) == 0 {
-		blkHeader := <-listener
+		select {
+		case err := <-subs.Err():
+			log.Error("Restarting L1 block monitoring. Errored with: %s", err)
+			listener, subs = a.ethClient.BlockListener()
 
-		// don't process blocks if the node is stopping
-		if atomic.LoadInt32(a.stopNodeInterrupt) == 1 {
-			return
-		}
+		case blkHeader := <-listener:
+			// don't process blocks if the node is stopping
+			if atomic.LoadInt32(a.stopNodeInterrupt) == 1 {
+				break
+			}
 
-		// ignore blocks if bootstrapping is happening
-		if atomic.LoadInt32(a.bootstrappingComplete) == 0 {
-			log.Trace("Node in bootstrap - ignoring block %s", blkHeader.Hash())
-			continue
-		}
+			// ignore blocks if bootstrapping is happening
+			if atomic.LoadInt32(a.bootstrappingComplete) == 0 {
+				log.Trace("Node in bootstrap - ignoring block %s", blkHeader.Hash())
+				continue
+			}
 
-		block, err := a.ethClient.BlockByHash(blkHeader.Hash())
-		if err != nil {
-			log.Panic("could not fetch block for hash %s. Cause: %s", blkHeader.Hash().String(), err)
-		}
-		blockParent, err := a.ethClient.BlockByHash(block.ParentHash())
-		if err != nil {
-			log.Panic("could not fetch block's parent with hash %s. Cause: %s", block.ParentHash().String(), err)
-		}
+			block, err := a.ethClient.BlockByHash(blkHeader.Hash())
+			if err != nil {
+				log.Panic("could not fetch block for hash %s. Cause: %s", blkHeader.Hash().String(), err)
+			}
+			blockParent, err := a.ethClient.BlockByHash(block.ParentHash())
+			if err != nil {
+				log.Panic("could not fetch block's parent with hash %s. Cause: %s", block.ParentHash().String(), err)
+			}
 
-		common.LogWithID(
-			a.shortID,
-			"Received a new block b_%d(%d)",
-			common.ShortHash(blkHeader.Hash()),
-			blkHeader.Number.Uint64(),
-		)
-		encodedBlock, err := common.EncodeBlock(block)
-		if err != nil {
-			log.Panic("could not encode block with hash %s. Cause: %s", block.Hash().String(), err)
-		}
-		encodedBlockParent, err := common.EncodeBlock(blockParent)
-		if err != nil {
-			log.Panic("could not encode block's parent with hash %s. Cause: %s", block.ParentHash().String(), err)
+			common.LogWithID(
+				a.shortID,
+				"Received a new block b_%d(%d)",
+				common.ShortHash(blkHeader.Hash()),
+				blkHeader.Number.Uint64(),
+			)
+			encodedBlock, err := common.EncodeBlock(block)
+			if err != nil {
+				log.Panic("could not encode block with hash %s. Cause: %s", block.Hash().String(), err)
+			}
+			encodedBlockParent, err := common.EncodeBlock(blockParent)
+			if err != nil {
+				log.Panic("could not encode block's parent with hash %s. Cause: %s", block.ParentHash().String(), err)
+			}
+			a.blockRPCCh <- blockAndParent{encodedBlock, encodedBlockParent}
 		}
-		a.blockRPCCh <- blockAndParent{encodedBlock, encodedBlockParent}
 	}
+
+	log.Info("Stopped monitoring for l1 blocks")
+	// make sure it cleanly unsubscribes
+	subs.Unsubscribe()
```

Comment:
Should this be in a defer before the select in case something in the main loop panics?

Reply:
Good point. I actually don't know how the defer will work because we might override the pointer 🤔 

I think it's fine because pretty much all bets are off when we panic.

Adding a TODO mentioning that though, for when we upstream the errors. WDYT ?

Reply:
yeah seems fair, if we panic in there I don't think we're recovering :)